### DISCUSSION TITLE
CSS: consider reuniting "Formal syntax" and "Values"

### DISCUSSION THREAD

**Discussion by wbamberg** [mdn/mdn-community#218] [Content]
In the old days, on CSS pages, the formal syntax section came before the "Values" section, inside "Syntax":

* 2014: https://web.archive.org/web/20140305011542/https://developer.mozilla.org/en-US/docs/Web/CSS/border#Syntax

The "Values" section described items listed in the formal syntax.

Because formal syntax was [hard to read](https://web.archive.org/web/20200319022331/https://developer.mozilla.org/en-US/docs/Web/CSS/border#Formal_syntax), we decided to move it further down the page in favour of more informal descriptions, first into a section just after "Values":

* 2020: https://web.archive.org/web/20200319022331/https://developer.mozilla.org/en-US/docs/Web/CSS/border#Syntax

...then eventually to its own top-level section, miles away from "Values":

* 2021: https://web.archive.org/web/20210413221123/https://developer.mozilla.org/en-US/docs/Web/CSS/border#formal_syntax

This causes coherency problems in our pages: the "Values" section still refers to things that are introduced in the formal syntax (e.g. `<line-width>` in the example above). In fact these two sections are quite closely linked.

I wonder if, now the formal syntax is easier to read, we should move it back before the "Values" section, so these two sections can work better again.

**Comment by wbamberg** [comment]
@dipikabh asked whether there was history about the names of the sections, since our current "Syntax" block is not really syntax definition as all, but more like examples of the syntax.

IIRC originally there was only what we now call "Formal syntax", and it was called just "Syntax": https://web.archive.org/web/20140305011542/https://developer.mozilla.org/en-US/docs/Web/CSS/border#Syntax. We wanted to have a more approachable/readable way to show the syntax, and added these sections giving example syntax. Then we needed a new name for the original syntax, so it got the name "formal syntax": https://web.archive.org/web/20180530112352/https://developer.mozilla.org/en-US/docs/Web/CSS/border#Syntax.

I still like the "example syntax" and think that probably for most people most of the time it's more likely to be useful than the formal syntax.

I think now we could perhaps have something like:

```
## Syntax

// syntax example

### Syntax definition (/Formal syntax)

// formal syntax

### Values

// values dl

```

**Comment by dipikabh** [comment]
Porting the slack discussion here for everybody's eyes and opinions:

@dipikabh:
- 'Syntax' should probably be called 'Syntax usage' (or another variation?) because we list examples in this section using the possible syntax options and not the real syntax.
- Is there a reason why 'Formal definition' is before 'Formal syntax'?
- I agree 'Values' should come after FS. Another point to consider is if 'Values' should be an H2 or H3.
- So could the section order be: Try It - Constituent properties (if applicable) - Description (if there is more to say than the intro para) - Syntax usage - FD - FS - Values - Examples - Spec - BCD - See also


@teoli2003:
- _Syntax usage_. What about Usage?
- I think we put FS after Examples because people are using Examples more. And being at the bottom of the page (because _Spec/BCD/See also_ are usually sort) still allows interested people to find it quickly.
- If we put FS up again, maybe we can hide it by default like in a `<summary>` /`<details>` combo. That way, it doesn't push examples so far away and is still one click away for people interested.
- There was a lot of user research about this back then. It may be interesting to conduct user interviews again..

@wbamberg:
 - FS is not after examples. And hiding it in `<details>` still creates the problem that "Values" does not make sense unless people expand it.

> Try It - Constituent properties (if applicable) - Description (if there is more to say than the intro para) - Syntax usage - FD - FS - Values - Examples - Spec - BCD - See also

I would like to consider grouping "syntax usage", "fs", "values" under an H2, to emphasise the fact that they are part of the same aspect of the documentation (and distinct from say BCD or Specs). And I'm not sure "Syntax usage" needs a heading. Also I think mockups help with this :slightly_smiling_face:

**Reply by dipikabh** [reply]
@wbamberg I like the idea of grouping "FS", "Values", "Usage". "FS" an H2 and "Values" and "Usage" H3s under it. I would retain the "Usage" heading just to have a pointer to access that section quickly. So a slightly different version from what Will has proposed above:

```
## Formal syntax
// formal syntax

### Values
// values dl

### Usage
// syntax example
```

- Where does FD fit in the scheme of things?
- I have seen "Description" in different places. I feel it should be right after "Try it".

**Reply by wbamberg** [reply]
> * Where does FD fit in the scheme of things?

Good question. I don't know yet :). I think when we get to webref-ify this, we might have a better idea.

> * I have seen "Description" in different places. I feel it should be right after "Try it".

It's fairly conventional in our docs for this to live after "Syntax", e.g.:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize

I like this for a couple of reasons:

- I think people are more likely to want to know syntax when they look up information in the reference, but are less likely to want to wade through the words (of course this is a user testing question). IME people only read words as a last resort.
- We sometimes have more prose sections for things like "Accessibility concerns" and it seems nice to glom all the prose-y things together.

(A meta-question here is whether we should do user testing to resolve all page structure questions in this thread, and whether we should block the specific change here (reuniting FS and Values) on that.)

> ```
> ## Formal syntax
> // formal syntax
> 
> ### Values
> // values dl
> 
> ### Usage
> // syntax example
> ```

I'd be OK with this, but would call it just "Syntax". And I still prefer having the syntax examples first, because I think it is more often of use to more people. (I use the formal syntax when I have to.)

**Comment by Rumyra** [comment]
> I think we put FS after Examples because people are using Examples more. And being at the bottom of the page (because Spec/BCD/See also are usually sort) still allows interested people to find it quickly.

I think this is an important point, I did a litmus test when we started considering webref for formal syntax instead of data. Most of the feedback was the same when asked whether users take note of the the formal syntax sections, like:

"Hardly ever! Can be pretty baffling on some longer entries."
"Can't say I ever noticed it"

That's not to say I'm against moving it as I think categorically it does fit better within the syntax and values parts. But maybe we should consider which parts of a CSS reference page are the most useful for our users (I assume this is why the original change was made)

**Reply by wbamberg** [reply]
> > I think we put FS after Examples because people are using Examples more. And being at the bottom of the page (because Spec/BCD/See also are usually sort) still allows interested people to find it quickly.
> 
> I think this is an important point, I did a litmus test when we started considering webref for formal syntax instead of data. Most of the feedback was the same when asked whether users take note of the the formal syntax sections, like:
> 
> "Hardly ever! Can be pretty baffling on some longer entries." "Can't say I ever noticed it"
> 
> That's not to say I'm against moving it as I think categorically it does fit better within the syntax and values parts. But maybe we should consider which parts of a CSS reference page are the most useful for our users (I assume this is why the original change was made)

Yes, people do say that. I hope it might be less baffling now, but it's never going to be easy, especially in more complicated cases.

What I don't get, though, is that the "Values" section does not make sense without the FS. For example, https://developer.mozilla.org/en-US/docs/Web/CSS/border#values - what is `<line-width>` supposed to be, here? So either people are somehow understanding this anyway(?), or they are not understanding this either, and they're just relying on copying the syntax-by-example bit or the interactive example.

But if the latter, then our pages aren't effectively describing the syntax of properties at all.

Perhaps we could do some more in-depth user testing to see how people do actually use these pages.

**Reply by rubiesonthesky** [reply]
> What I don't get, though, is that the "Values" section does not make sense without the FS. For example, https://developer.mozilla.org/en-US/docs/Web/CSS/border#values - what is `<line-width>` supposed to be, here? So either people are somehow understanding this anyway(?), or they are not understanding this either, and they're just relying on copying the syntax-by-example bit or the interactive example

I think some users already know some CSS and / or these these terms and can understand that these are referring the examples. 

I think for this page, it's little bit confusing to have  consistuent properties and values so far apart.

Reading formal syntax or formal definition feels like chore. They are not easy to skim through, especially in mobile. I think just making font size a little bit smaller in formal syntax would help, as it would then fit better on screen. If i need to scroll to see whole syntax (as is the case for border), I'm going to refer some other source.

This is just one user's perspective for this. There may not be one easy and simple answer for this question.

**Comment by dipikabh** [comment]
Based on the sections that are most helpful to readers, how about the order somewhat along the lines of what's described here. This proposal does not take into account if there'll be any implementation hiccups (in content or platform), just throwing ideas out there. Also, not sure if this will be extendable to JS docs.

- In the "Usage" section (proposed new name for the "Syntax section" but no consensus yet), we can state the syntax in a comment before we add a usage sample. This would solve the problem that "Values" section uses parameter names that are only referred to later in the "Formal syntax" section.
- We can also try to limit the length of the "Usage" section to show only a sample of each kind.
- Should "Values" be called "Parameters" or "Options" instead? I am suggesting below to turn it into an H2.
- How about moving FS and FD under "Specifications" because that's where they are coming from? Also, how about making FS and FD collapsible so that readers expand the sections only if they need to but the information is there.

Using https://developer.mozilla.org/en-US/docs/Web/CSS/animation-delay#syntax as an example here. 

```md
## Try it
// interactive example

## Usage
/* For a single animation */
/* animation-delay: <time>; */
animation-delay: 3s;
animation-delay: -1500ms;

/* For multiple animations */
/* animation-delay: <time>, <time>; */
animation-delay: 2.1s, 480ms;

## Parameters
// dl

## Description

## Examples

## Specifications
// url table
### Formal syntax
### Formal definition

## Browser compatibility

## See also

```

Consider another page https://developer.mozilla.org/en-US/docs/Web/CSS/border-style#syntax.
- The "Syntax" section should not list all the possible values; that should be done in the "Values" section.
- Perhaps even the list of all applicable "Global values" should be moved out of "Syntax" and mentioned in the "Values" section.

```md
## Try it
// interactive example

## Constituent properties

## Usage
/* 1 value applies to all sides */
/* border-style: <line-style>; */ 
border-style: dotted;

/* 3 values apply to top | left and right | bottom */
/* border-style: <line-style> <line-style> <line-style>; */ 
border-style: hidden double dashed;

...

## Parameters
// dl
// also mention the global values that this property can be assigned

## Description

## Examples

## Specifications
// url table
### Formal syntax
### Formal definition

## Browser compatibility

## See also

```

**Reply by wbamberg** [reply]
A few things...

- currently we have a very consistent set of sections at the back end of a page: Examples/Specifications/Browser compatibility/See also. This is consistent not just in CSS but across most of our reference docs. I'm not a fan of making the CSS docs different here (as we would do if we put this stuff in "Specifications").
- I don't really think it belongs in "Specifications" . Yes, it comes out of the specs, but so does almost all the page.
- To me "Parameters" is not a good fit here. But maybe this is just subjective.
- I think "Usage" is quite generic compared with "Syntax". For example, "Usage" could be a prose section describing how to use a thing. "Syntax" is specifically describing the structure of the property.

> we can state the syntax in a comment before we add a usage sample. This would solve the problem that "Values" section uses parameter names that are only referred to later in the "Formal syntax" section.

Maybe. Syntax is complicated sometimes and hard to express precisely in prose. This also doesn't solve the other side of the problem, that the formal syntax lists values that need an explanation: for example, in https://developer.mozilla.org/en-US/docs/Web/CSS/list-style-type#formal_syntax we should explain what `<counter-style-name>` is.

> Also, how about making FS and FD collapsible so that readers expand the sections only if they need to but the information is there.

Yes, people have suggested this in the past. I don't really like it because I think that if you hide things people won't find them, including people who could benefit from them. Really you almost might as well drop it, if you don't think it is useful enough to show.

I have changed my mind about formal syntax. I used to think it was not useful, but now I think it is very useful, but poorly presented, which is why I tried to make it better presented. I think with even just a little learning on the part of readers it is a very effective way to communicate syntax, much quicker and more precise than prose. I appreciate that many people don't think it's useful. I hope that this might change now we are better at presenting it, and especially if we can integrate it more coherently into the page. I think that it's a good idea to split "syntax" into an informal syntax-by-example, which works most of the time for most people, and a formal syntax which is much more precise and covers all the cases.

**Reply by teoli2003** [reply]
>     * To me "Parameters" is not a good fit here. But maybe this is just subjective.
Yes, parameters is not the right word. The proper word is _Value_, at least for properties.
![image](https://user-images.githubusercontent.com/1466293/188441586-ec03a388-873e-4d73-9c98-0c0b9de12787.png)

A keyword is a specific kind of values.

**Reply by dipikabh** [reply]
All fair points, @wbamberg and thanks for providing the historical context.

Irrespective of where the "Values" section ultimately sits:
- Could we change "Values" from an H3 to an H2 title?
- The "Values" section should use argument names as they appear in "Formal syntax". For example, in https://developer.mozilla.org/en-US/docs/Web/CSS/list-style-type#syntax, the "Values" section mentions `custom-indent` but there is no `custom-indent` in Formal syntax. 
-  In the "Syntax" section, I proposed adding a part of syntax in a comment, which is specific to that syntax example. This comment will use the argument names that are explained in the "Values" section. I feel this is one way to connect "Formal syntax" with "Values" with "Syntax".

As an example (partial), we could do the following in case of https://developer.mozilla.org/en-US/docs/Web/CSS/list-style-type#syntax:

```
## Syntax
/* list-style-type: <counter-style-name> */
list-style-type: disc;
list-style-type: circle;

/* list-style-type: <string> */
list-style-type: '-';

## Values
<counter-style-name>
    def
<symbols()>
   def
<string>
   def
```

**Reply by teoli2003** [reply]
> All fair points, @wbamberg and thanks for providing the historical context.

>     * The "Values" section should use argument names as they appear in "Formal syntax". For example, in https://developer.mozilla.org/en-US/docs/Web/CSS/list-style-type#syntax, the "Values" section mentions `custom-indent` but there is no `custom-indent` in Formal syntax.

I think one of the confusion come that things like `custom-indent`aren't argument names, they are type names, that even could appear several times.

E.g. (invented):   `my-property: <custom-ident> <custom-ident> <length> <length>?` describes either:
- a three-value syntax, where refering to `<custom-ident>` without context is ambiguous as there are 2 custom-ident values.
- a four-value syntax, with two `<custom-ident>` and two `<length-value>`.


> 
> As an example (partial), we could do the following in case of https://developer.mozilla.org/en-US/docs/Web/CSS/list-style-type#syntax:
> 
> ```
> ## Syntax
> /* list-style-type: <counter-style-name> */
> list-style-type: disc;
> list-style-type: circle;
> 
> /* list-style-type: <string> */
> list-style-type: '-';
> 
> ## Values
> <counter-style-name>
>     def
> <symbols()>
>    def
> <string>
>    def
> ```
This work when these are "or" values (`|`), but this is far from being the general case.

In many cases, we have 2-value, 3-value, and 4-value syntaxes with different meanings for value #2 depending on the chosen syntax.
E.g [margin](https://developer.mozilla.org/en-US/docs/Web/CSS/margin) where the second value is always of the same type but has a different meaning, depending on the syntax.

Or try describing the `linear-gradient` syntax…

**Reply by dipikabh** [reply]
My idea was to add just one more line as a comment to the syntax example - to express a specific syntax example as its equivalent syntax (not the entire syntax) using the type names described in "Values". For example, in the previous [comment](https://github.com/mdn/mdn-community/discussions/218#discussioncomment-3526088), I tried to tackle https://developer.mozilla.org/en-US/docs/Web/CSS/border-style#syntax:

```
/* 3 values apply to top | left and right | bottom */
/* border-style: <line-style> <line-style> <line-style>; */ 
border-style: hidden double dashed;
```

This example refers to different sides depending on the number of values but the type name is `<line-style>` and it appears by the same name in "Values" and "Formal syntax". The `<line-style>` name in "Values" will now also have a reference in the "Syntax" section as well.

Similarly, there's an example of the multiple occurring value of <time> in the earlier comment:
```
/* For multiple animations */
/* animation-delay: <time>, <time>; */
animation-delay: 2.1s, 480ms;
```
and `<time>` can then be explained in the "Values" section (and it also occurs in Formal syntax). So it would be less problematic now even if Formal syntax is few sections away.

(They are not arguments or parameters. Sorry, I am getting stuck on how to refer to different parts in a syntax. The name `<custom-indent>` in "Values" should ideally be something that appears in "Formal syntax", just like "Values" refers to `<symbols()>` and `<string>` by the same names as they appear in "Formal syntax" https://developer.mozilla.org/en-US/docs/Web/CSS/list-style-type#values.)

**Reply by teoli2003** [reply]
Note that these extra lines of "formal syntax" are different from the ones obtained from  w3c/webref and will have to be crafted manually.

**Reply by dipikabh** [reply]
Yes, that's correct, it will be a manual effort. It will be a process change where we not only provide a syntax example but also refer to the values used in that example.

I came here to report something I just spotted while browsing pseudo-classes. Interestingly, there are some pages that use the title "Parameters" in place of "Values":
- https://developer.mozilla.org/en-US/docs/Web/CSS/:dir#parameters
- https://developer.mozilla.org/en-US/docs/Web/CSS/filter-function/blur#parameters (and a lot of more pages under `filter-function`)
- https://developer.mozilla.org/en-US/docs/Web/CSS/path#parameters

Perhaps they all need to be fixed for consistency.

**Reply by teoli2003** [reply]
They are not values but parameters either of the functions (like `blur(some-parameters)`) or of the pseudo-class (like `:dir(ltr)`). Only properties have values.

In CSS, pseudo-classes, pseudo-elements, properties, functions, etc. are different structures. They will have a different _page-type_. Here are some of the proposed page types for CSS (in addition of _guide_):


    css-at-rule-descriptor
    css-at-rule
    css-data-type
    css-function
    css-keyword
    css-media-feature
    css-property
    css-selector
    css-shorthand-property

More [details](https://github.com/mdn/content/discussions/5162#discussion-3374795).

**Reply by wbamberg** [reply]
> 
> Irrespective of where the "Values" section ultimately sits:
> 
> * Could we change "Values" from an H3 to an H2 title?

The reason I suggested having it as an H3 is that we can use the structure to show how "syntax by example", "formal syntax" and "values" are all closely linked and depend on each other, unlike H2's such as BCD that seem to stand alone. If we make "Values" an H2 we're not able to communicate that.

Is there a reason for making it an H2?

> * The "Values" section should use argument names as they appear in "Formal syntax". For example, in https://developer.mozilla.org/en-US/docs/Web/CSS/list-style-type#syntax, the "Values" section mentions `custom-indent` but there is no `custom-indent` in Formal syntax.

Yes, absolutely. This is an error (fundamentally, it's because webref doesn't expand `<counter-style-name>` as a `<custom-ident>`, I think because it isn't exactly a `<custom-ident>`: as the spec says: ["`<counter-style-name>` is a `<custom-ident>` that is not an ASCII case-insensitive match for none"](https://drafts.csswg.org/css-counter-styles/#typedef-counter-style-name).

> * In the "Syntax" section, I proposed adding a part of syntax in a comment, which is specific to that syntax example. This comment will use the argument names that are explained in the "Values" section. I feel this is one way to connect "Formal syntax" with "Values" with "Syntax".
> 
> 
> As an example (partial), we could do the following in case of https://developer.mozilla.org/en-US/docs/Web/CSS/list-style-type#syntax:
> 

I think this is worth a try :). I'm worried that it will get very complicated in some cases, and the comment in the example will end up looking like an ad hoc version of the formal syntax. So it might be worth trying to write this out for a selection of more or less complicated properties, and seeing how it feels.

I still think it would be good, for coherence, to keep the formal syntax close to "values" rather than banishing it to some far-off bit of the page.

**Reply by dipikabh** [reply]
> Is there a reason for making it an H2?

Only if "Syntax" and Formal syntax" stay H2.

I very much like the proposal of grouping "syntax by example", "formal syntax", and "values" under an H2 (still can't think of what we'll call this H2 though). The only problem I foresee is that as H3s, these will not be directly accessible from the page TOC at the current moment.

How about the order "syntax by example", "values", "formal syntax" though because "values" does help in deciphering some of the syntax use cases? Would FD stay out of the trio?

So would it be:
Try It - Constituent properties (if applicable) - H2 ( H3:Syntax by example - H3:Values - H3:FS) - Description - FD - Examples - Accessibility concerns (if applicable) - Spec - BCD - See also

> I think this is worth a try :). I'm worried that it will get very complicated in some cases, and the comment in the example will end up looking like an ad hoc version of the formal syntax. So it might be worth trying to write this out for a selection of more or less complicated properties, and seeing how it feels.

IMO if a syntax example/use case exists, then it is following a specific syntax. So "theoretically", we should be able to break down every syntax use case in terms of the syntax and values it is using. We should definitely try out the most complex cases 👍. (I can have a go at some of them.)

The other simpler solution is to change the title "Syntax" to "Syntax by example" or "Syntax uses cases" (and I guess this would be only for certain page types) and also ensure (over time) that "Values" uses the same value names as they appear in "Formal syntax".

(As a good practice and to ensure "syntax by example", "formal syntax", and "values" all tie in with each other, I'd recommend that over time we add that comment in Syntax alongside the use case – this, of course, can be an initiative separate from the objective of this discussion.)

**Reply by wbamberg** [reply]
> > Is there a reason for making it an H2?
> 
> Only if "Syntax" and Formal syntax" stay H2.
> 
> I very much like the proposal of grouping "syntax by example", "formal syntax", and "values" under an H2 (still can't think of what we'll call this H2 though). The only problem I foresee is that as H3s, these will not be directly accessible from the page TOC at the current moment.

I wondered if that was the reason, but if so I think we should do what's right for the page structure now, and hope the ToC catches up soon (do you know where that work is/whether something is blocking it?)

> How about the order "syntax by example", "values", "formal syntax" though because "values" does help in deciphering some of the syntax use cases? Would FD stay out of the trio?
> 
> So would it be: Try It - Constituent properties (if applicable) - H2 ( H3:Syntax by example - H3:Values - H3:FS) - Description - FD - Examples - Accessibility concerns (if applicable) - Spec - BCD - See also
> 

I think this would be fine as long as we could get the values listed in the "syntax by example" block. Personally I don't think we need an addition H3 for the "syntax by example" block, I think it would be fine to have:

```
## Syntax

// syntax by example

### Values

### Formal syntax
```

>  Would FD stay out of the trio?

Yes. I don't know how that would end up. I think we will maybe change it a bit when we look at powering it through webref instead iof mdn/data, but suggest leaving it where it is for now.

> > I think this is worth a try :). I'm worried that it will get very complicated in some cases, and the comment in the example will end up looking like an ad hoc version of the formal syntax. So it might be worth trying to write this out for a selection of more or less complicated properties, and seeing how it feels.
> 
> IMO if a syntax example/use case exists, then it is following a specific syntax. So "theoretically", we should be able to break down every syntax use case in terms of the syntax and values it is using. We should definitely try out the most complex cases 👍. (I can have a go at some of them.)

Theoretically yes, for sure. It's more a question of whether the block gets so long and complex that you are better off with the formal syntax.

> The other simpler solution is to change the title "Syntax" to "Syntax by example" or "Syntax uses cases" (and I guess this would be only for certain page types) and also ensure (over time) that "Values" uses the same value names as they appear in "Formal syntax".

I think it is important for coherence that items in "Values" have the same names as items in formal syntax.

Maybe I should start a PR with a few pages in, so we have something concrete to work on?

**Reply by teoli2003** [reply]
> I think it is important for coherence that items in "Values" have the same names as items in formal syntax.

Do we have a way to detect when a formal syntax in webref changed? (This happens from time to time) because we need to update the page on MDN.

**Reply by wbamberg** [reply]
> > I think it is important for coherence that items in "Values" have the same names as items in formal syntax.
> 
> Do we have a way to detect when a formal syntax in webref changed? (This happens from time to time) because we need to update the page on MDN.

No. But I'm not sure that's really different from any other way we don't get notifications when the spec changes in ways that should trigger MDN updates.

**Reply by teoli2003** [reply]
It is fairly different. A formal syntax change can make a page incoherent, while another change, it can make the change outdated, and only if a browser implements the change.

**Reply by dipikabh** [reply]
> I think we should do what's right for the page structure now, and hope the ToC catches up soon (do you know where that work is/whether something is blocking it?)

A [discussion](https://github.com/mdn/mdn-community/discussions/162) is open and the engineering team is aware of it. I am not aware of the implementation timeline.

> Maybe I should start a PR with a few pages in, so we have something concrete to work on?

Sounds good!

**Reply by Rumyra** [reply]
@dipikabh I suggest you raise an issue on yari to expose the h3s in TOC - I think we're mainly in consensus this is a good thing - we shouldn't block making sure page structures are there first, because the opposite is also true, we wait for page structures before doing it - but either or are going to influence moving forward

**Comment by yarusome** [comment]
The crucial point here is that different reader groups have different expectations for the "Syntax" section. There're at least two groups:

  1. beginner- to intermediate-level devs that try to grasp the basic ideas from this section;
  2. experienced devs that look up this section for assurance and details.

Some threads here are focusing on a single group, but it's better to accommodate both.

Another issue I found during the work on CSS Color Module is that the current structure makes it awkward to explain functional notations, e.g. starting the "Values" section with a note on the syntax as on the [`hsl()`](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/hsl#values) page.

Therefore, I think the proposed structures could be further tweaked like this:

```md
## Syntax

[{Optional} Intuitive rephrasing of the syntax here][1]

[Current "Formal syntax" here][2]

### Values

### "Examples" (<- A bad name for a placeholder, current "Syntax" here)

## Definition

[Current "Formal definition" here]
```

[1] serves Group 1 and functionnal notations, and [2] serves Group 2. However, there're some caveats:

  1. There should be a clear distinction between [1] and [2] - either a phrasing or a visual one - so that readers won't be confused as "Why there're two different syntaxes?".
  2. As already mentioned in other threads, the current implementation of `{{CSSSyntax}}` sometimes makes *really lengthy* expansions, e.g. [`display`](https://developer.mozilla.org/en-US/docs/Web/CSS/display#formal_syntax) and [`transition-timing-function`](https://developer.mozilla.org/en-US/docs/Web/CSS/transition-timing-function#formal_syntax), which make readers lost in details. (Btw, this macro has been needing updating for quite some time, see https://github.com/mdn/yari/pull/8909.)

The culprits for Point 2 are:

  - The macro doesn't stop at data types already exist on MDN, which is due to the ad hoc logic on [Line 16 to 17](https://github.com/mdn/yari/blob/main/kumascript/src/lib/css-syntax.ts#L16-L17).
  - The format is rather spatially uneconomical. For example, the syntax of [`<display-internal>`](https://developer.mozilla.org/en-US/docs/Web/CSS/display-internal#formal_syntax) could use flexbox or `display: inline-block` (see [CSS Text](https://drafts.csswg.org/css-text/#line-break-details)) to save space *and* align the items at the same time.

In case the formal syntax is still too lengthy after the above optimizations, a last resort is to set `max-height` on this part.

---

Incidentally, I'm working on an automation script to diff mdn/data and webref so that future updates to mdn/data could be much easier.

---

A demo with `white-space: pre-wrap` + `display: inline-block` for [`<display-internal>`](https://developer.mozilla.org/en-US/docs/Web/CSS/display-internal#formal_syntax):

![screenshot](https://github.com/mdn/mdn-community/assets/97945148/c0024ac4-0a13-4f1e-a449-afe1d4997ac8)

Code:

```html
<pre class="notranslate" style="white-space: pre-wrap;"><span class="token property" id="<display-internal>">&lt;display-internal&gt; = </span><br><span style="display: inline-block;">  <span class="token keyword">table-row-group</span>      <a href="/en-US/docs/Web/CSS/Value_definition_syntax#single_bar" title="Single bar: exactly one of the entities must be present">|</a></span><span style="display: inline-block;">  <span class="token keyword">table-header-group</span>   <a href="/en-US/docs/Web/CSS/Value_definition_syntax#single_bar" title="Single bar: exactly one of the entities must be present">|</a></span><span style="display: inline-block;">  <span class="token keyword">table-footer-group</span>   <a href="/en-US/docs/Web/CSS/Value_definition_syntax#single_bar" title="Single bar: exactly one of the entities must be present">|</a></span><span style="display: inline-block;">  <span class="token keyword">table-row</span>            <a href="/en-US/docs/Web/CSS/Value_definition_syntax#single_bar" title="Single bar: exactly one of the entities must be present">|</a></span><span style="display: inline-block;">  <span class="token keyword">table-cell</span>           <a href="/en-US/docs/Web/CSS/Value_definition_syntax#single_bar" title="Single bar: exactly one of the entities must be present">|</a></span><span style="display: inline-block;">  <span class="token keyword">table-column-group</span>   <a href="/en-US/docs/Web/CSS/Value_definition_syntax#single_bar" title="Single bar: exactly one of the entities must be present">|</a></span><span style="display: inline-block;">  <span class="token keyword">table-column</span>         <a href="/en-US/docs/Web/CSS/Value_definition_syntax#single_bar" title="Single bar: exactly one of the entities must be present">|</a></span><span style="display: inline-block;">  <span class="token keyword">table-caption</span>        <a href="/en-US/docs/Web/CSS/Value_definition_syntax#single_bar" title="Single bar: exactly one of the entities must be present">|</a></span><span style="display: inline-block;">  <span class="token keyword">ruby-base</span>            <a href="/en-US/docs/Web/CSS/Value_definition_syntax#single_bar" title="Single bar: exactly one of the entities must be present">|</a></span><span style="display: inline-block;">  <span class="token keyword">ruby-text</span>            <a href="/en-US/docs/Web/CSS/Value_definition_syntax#single_bar" title="Single bar: exactly one of the entities must be present">|</a></span><span style="display: inline-block;">  <span class="token keyword">ruby-base-container</span>  <a href="/en-US/docs/Web/CSS/Value_definition_syntax#single_bar" title="Single bar: exactly one of the entities must be present">|</a></span><span style="display: inline-block;">  <span class="token keyword">ruby-text-container</span>  </span><br></pre>
```

Given the above format, if extra-long items are allowed to occupy more than one column, there arises an interesting math problem to decide the most economical column width.

**Reply by wbamberg** [reply]
> The macro doesn't stop at data types already exist on MDN, which is due to the ad hoc logic on [Line 16 to 17](https://github.com/mdn/yari/blob/main/kumascript/src/lib/css-syntax.ts#L16-L17).

Indeed. But [this was a deliberate choice](https://github.com/mdn/yari/pull/6871), because making all the syntaxes stop when there is the possibility of linking to another page makes the formal syntax very fragmentary, so people would have to navigate to lots of pages in order to get a sense of what's possible for a given property. `display` is a good example of that actually: yes it is long(ish), but if it delegated to all the types it would be very hard to get a sense of what the possible concrete values are.

The ad hoc treatment of `<gradient>` and `<color>` are because expanding those types really does make the syntax huge, and they end up in many syntaxes.

> The format is rather spatially uneconomical. For example, the syntax of [<display-internal>](https://developer.mozilla.org/en-US/docs/Web/CSS/display-internal#formal_syntax) could use flexbox or display: inline-block (see [CSS Text](https://drafts.csswg.org/css-text/#line-break-details)) to save space and align the items at the same time.

I think this is a tradeoff: you save vertical space, but IMO it's much easier to scan a 1-column list than multiple columns. 🤷 .
